src/common/config.cc

```diff
@@ -267,8 +263,7 @@ int md_config_t::parse_config_files_impl(const std::list<std::string> &conf_file
      * If cluster name is not set yet, use the prefix of the
      * basename of configuration file as cluster name.
      */
-    const char *fn = c->c_str();
-    std::string name(basename(fn));
+    std::string name = c->substr(c->find_last_of("/") + 1);
```

Comment:
On 23-6-2016 08:50, Kefu Chai wrote:
> In src/common/config.cc
> <https://github.com/ceph/ceph/pull/9845#discussion_r68183228>:
> 
>> @@ -267,8 +263,7 @@ int md_config_t::parse_config_files_impl(const std::list<std::string> &conf_file
>>       * If cluster name is not set yet, use the prefix of the
>>       * basename of configuration file as cluster name.
>>       */
>> -    const char *fn = c->c_str();
>> -    std::string name(basename(fn));
>> +    std::string name = c->substr(c->find_last_of("/") + 1);
> 
> as |c_str()| returns a "const char/", and per POSIX standard[1],
> |basename()| expects a "char/", and it might change the string passed
> in. even the freebsd and glibc kindly offers |basename(const char*)|
> implementation.

Yup,
That is where all my hassle started: |const char|.
FreeBSD does grok that, but only if the right library is included.
At the time I looked at several man-pages, but not at the POSIX std.

Note that performance is is not an issue.
So if manual crafted code needs to do it that is fine.

Not sure why it went from manual code to to basename() first place.

--WjW

> so, yeah, i agree that we do need to change for standard compliance and
> better portability.
> 
> maybe we can have
> 
> auto start = c->rfind('/') + 1;
> auto end = c->find(".conf", start);
> if (end == name.npos) {
>   ...
> } else {
>   cluster = name.substr(start, end - start);
> }
> ---
> [1] http://pubs.opengroup.org/onlinepubs/9699919799/
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/9845/files/2757cdc16c40a9883556423df4ab2f4637285f20#r68183228>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APVj-jFL75XKmOvnpGOV_m5eUUlN2wubks5qOizHgaJpZM4I6wKj>.
>